JAVA-2000: Fix channel shutdown when abortAllInFlight processes a Set…

olim7t opened the pull request:
…KeyspaceRequest

olim7t commented:
I tried to cover it with a unit test but it would rely on hashmap ordering (I think there need to be other callbacks in the map after the `SetKeyspace` one) and log parsing since the error is only shown at DEBUG. Overall I'm not sure it's worth it.

GregBestland commented:
Unfortunately I still see concurrent modification exception occuring even with this branch installed and used.

10:19:14.314 [s18-io-1] DEBUG i.n.c.AbstractChannelHandlerContext - An exception java.util.ConcurrentModificationException
	at com.datastax.oss.driver.shaded.guava.common.collect.HashBiMap$Itr.hasNext(HashBiMap.java:411)
	at com.datastax.oss.driver.internal.core.channel.InFlightHandler.abortAllInFlight(InFlightHandler.java:357)
	at com.datastax.oss.driver.internal.core.channel.InFlightHandler.abortAllInFlight(InFlightHandler.java:345)
	at com.datastax.oss.driver.internal.core.channel.InFlightHandler.exceptionCaught(InFlightHandler.java:293)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:285)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:264)
	at io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:256)
	at io.netty.channel.ChannelInboundHandlerAdapter.exceptionCaught(ChannelInboundHandlerAdapter.java:131)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:285)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:264)
	at io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:256)
	at io.netty.channel.ChannelHandlerAdapter.exceptionCaught(ChannelHandlerAdapter.java:87)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:285)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:264)
	at io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:256)
	at io.netty.handler.ssl.SslHandler.exceptionCaught(SslHandler.java:1048)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:285)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:264)
	at io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:256)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.exceptionCaught(DefaultChannelPipeline.java:1401)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:285)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:264)
	at io.netty.channel.DefaultChannelPipeline.fireExceptionCaught(DefaultChannelPipeline.java:953)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.handleReadException(AbstractNioByteChannel.java:125)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:174)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:646)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:581)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:498)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:460)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:884)
	at java.lang.Thread.run(Thread.java:748)
was thrown by a user handler's exceptionCaught() method while handling the following exception:
java.io.IOException: Connection reset by peer
	at sun.nio.ch.FileDispatcherImpl.read0(Native Method)
	at sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39)
	at sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:223)
	at sun.nio.ch.IOUtil.read(IOUtil.java:192)
	at sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:380)
	at io.netty.buffer.PooledUnsafeDirectByteBuf.setBytes(PooledUnsafeDirectByteBuf.java:288)
	at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1108)
	at io.netty.channel.socket.nio.NioSocketChannel.doReadBytes(NioSocketChannel.java:345)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:148)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:646)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:581)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:498)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:460)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:884)
	at java.lang.Thread.run(Thread.java:748)

It seem suspicious to me that with that boolean introduction that it would still be so reliable to reproduce, as the window that this should be able to happen has shrunk considerably.

GregBestland commented on core/src/main/java/com/datastax/oss/driver/internal/core/channel/InFlightHandler.java:
```diff
@@ -327,8 +328,13 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+    } else {
+      responseCallback = orphaned.remove(streamId);
+    }
     orphanedSize = orphaned.size();
```
This may not be the right way to handle this, but the problem I saw after the initial fix was that inflight.remove was being invoked while abort all inflight was being executed via the exceptionCaught(InFlightHandler.java:293) path.

olim7t committed:
JAVA-2000: Fix ConcurrentModificationException during channel shutdown

Motivation:

When a channel is shut down abruptly we iterate the inFlight map in
abortAllInFlight() to fail all pending queries. But in some cases,
failing a query will also indirectly mutate the map, causing a
ConcurrentModificationException.

We ran into this particular scenario:
- the channel initializes and the write of the initial STARTUP query is
  scheduled (but the write future is not complete yet)
- during the actual write task, an IOException is thrown ("connection
  reset by peer"). exceptionCaught() is called and invokes
  abortAllInFlight().
- abortAllInFlight() calls onFailure() on the
  ProtocolInitHandler.InitRequest corresponding to the STARTUP request.
  That calls ConnectInitHandler.setConnectFailure(), which closes the
  channel.
- closing the channel fails the write future of the STARTUP query, which
  invokes the write listener synchronously. The listener calls release()
  which removes the callback from inFlight.
- the initial iteration resumes and finds out that the map was modified.

Similar issues could happen if one of the aborted requests is a
SetKeyspaceRequest that calls abortAllInFlight() recursively.

Modifications:

- don't iterate inFlight directly, make a copy to avoid concurrent
  modifications.
- clear it immediately so that recursive invocations of
  abortAllInFlight() have no effect.
- ensure release() is lenient if the callback is not in inFlight. For
  clarity, also change it to never return the callback, the caller has
  to retrieve it itself (as already done in channelRead).

Result:

The initial call to abortInFlight() clear inFlight and fails the STARTUP
query. When the write listener invokes release(), that's a no-op because
the callback is not in the map anymore.

Co-authored-by: Greg Bestland <Greg.Bestland@datastax.com>

olim7t commented:
OK I've rearranged things a bit:
- clear the map first and iterate on a copy instead of using an extra boolean
- simply do nothing in `release()` -- which now sees an empty map. It also doesn't return the callback anymore, the caller has to look it up explicitly, which is a bit more clear.

I don't see the issue anymore in my local environment.
